ml-agents/mlagents/trainers/components/reward_signals/curiosity/signal.py

```diff
@@ -0,0 +1,172 @@
+        :return: Dict of stats that should be reported to Tensorboard.
+        """
+        forward_total, inverse_total = [], []
+        for _ in range(self.num_epoch):
+            update_buffer.shuffle()
+            buffer = update_buffer
+            for l in range(len(update_buffer["actions"]) // num_sequences):
+                start = l * num_sequences
+                end = (l + 1) * num_sequences
+                run_out_curio = self._update_batch(
+                    buffer.make_mini_batch(start, end), num_sequences
+                )
+                inverse_total.append(run_out_curio["inverse_loss"])
+                forward_total.append(run_out_curio["forward_loss"])
+
+        update_stats = {
+            "Losses/Curiosity Forward Loss": np.mean(forward_total),
+            "Losses/Curiosity Inverse Loss": np.mean(inverse_total),
+        }
+        return update_stats
+
+    def _update_batch(self, mini_batch, num_sequences):
+        """
+        Updates model using buffer.
+        :param num_sequences: Number of trajectories in batch.
+        :param mini_batch: Experience batch.
+        :return: Output from update process.
+        """
+        feed_dict = {
+            self.policy.model.batch_size: num_sequences,
+            self.policy.model.sequence_length: self.policy.sequence_length,
+            self.policy.model.mask_input: mini_batch["masks"].flatten(),
+            self.policy.model.advantage: mini_batch["advantages"].reshape([-1, 1]),
+            self.policy.model.all_old_log_probs: mini_batch["action_probs"].reshape(
+                [-1, sum(self.policy.model.act_size)]
+            ),
+        }
+        if self.policy.use_continuous_act:
+            feed_dict[self.policy.model.output_pre] = mini_batch["actions_pre"].reshape(
+                [-1, self.policy.model.act_size[0]]
+            )
+            feed_dict[self.policy.model.epsilon] = mini_batch[
+                "random_normal_epsilon"
+            ].reshape([-1, self.policy.model.act_size[0]])
+        else:
+            feed_dict[self.policy.model.action_holder] = mini_batch["actions"].reshape(
+                [-1, len(self.policy.model.act_size)]
+            )
+            if self.policy.use_recurrent:
+                feed_dict[self.policy.model.prev_action] = mini_batch[
+                    "prev_action"
+                ].reshape([-1, len(self.policy.model.act_size)])
+            feed_dict[self.policy.model.action_masks] = mini_batch[
+                "action_mask"
+            ].reshape([-1, sum(self.policy.brain.vector_action_space_size)])
+        if self.policy.use_vec_obs:
+            feed_dict[self.policy.model.vector_in] = mini_batch["vector_obs"].reshape(
+                [-1, self.policy.vec_obs_size]
+            )
+            feed_dict[self.model.next_vector_in] = mini_batch["next_vector_in"].reshape(
+                [-1, self.policy.vec_obs_size]
+            )
+        if self.policy.model.vis_obs_size > 0:
```

Comment:
Am I missing something (probably) or can you rewrite this with one for-loop?
``` python
if self.policy.model.vis_obs_size > 0:
    for i, _ in enumerate(self.policy.model.visual_in):
        _obs = mini_batch["visual_obs%d" % i]
        _next_obs = mini_batch["next_visual_obs%d" % i]
        if self.policy.sequence_length > 1 and self.policy.use_recurrent:
            (_batch, _seq, _w, _h, _c) = _obs.shape
            feed_dict[self.policy.model.visual_in[i]] = _obs.reshape(
                [-1, _w, _h, _c]
            )
            (_batch, _seq, _w, _h, _c) = _next_obs.shape
            feed_dict[self.model.next_visual_in[i]] = _next_obs.reshape(
                [-1, _w, _h, _c]
            )
        else:
            feed_dict[self.policy.model.visual_in[i]] = _obs
            feed_dict[self.model.next_visual_in[i]] = _next_obs
```

Reply:
I think you're right given the fact that the for loops and if statements are identical. 

This logic is in many places throughout the code - maybe we can go ahead and change them. Let me test and make sure it doesn't break anything here.

Reply:
OK, if you don't think it's an improvement or don't want to rewrite the rest of the code, don't sweat it.